### PR TITLE
Teleporter registry e2e test

### LINKED ISSUE #221: Integration Tests for Registry Updates
**Context and scope**
Now that off-chain Warp message support has been added to `coreth` and `subnet-evm` ([PR here](https://github.com/ava-labs/subnet-evm/pull/1028)), we should include Ginkgo integration tests of the relevant registry update flows.

**Discussion and alternatives**
Possible test cases include:
- Deploying new `TeleporterMessenger` version by modifying the Nick's method transaction such that the same contract byte code will be deployed to a different address.
- Adding the new contract address to the registry on one chain but not another. Check that messages are sent using the new version, but can't be received. 
- Check that messages sent from the other chain still use the old version and are able to be received.
- Add the contract address on the second chain, check that previously sent messages are now able to be received.
- Update the minimum required version for a dApp, ensure messages sent using the older version are no longer able to be received.

### PR DESCRIPTION
**Pull request by minghinmatthewlam**
## Why this should be merged
Adds an e2e test for testing the `TeleporterRegistry`, fixes #221.

## How this works
Tests:
- Setting Warp off-chain messages in the chain config and restarting nodes
- Adding new protocol version to `TeleporterRegistry` with signed Warp off-chain message
- Updating `minTeleporterVersion` of a dapp that uses the registry
- Makes sure receiving fails if Teleporter version is not `>= minTeleporterVersion`
- When both chains have the new deployed Teleporter, but only subnet B has it registered in registry. If we send from chain with registered new version, message will deliver to destination Teleporter, but message execution fails and destination dapp does not receive.
  - After subnet A registers new version, subnet B can deliver to dapp on subnet A again with newer Teleporter version

## How this was tested
Steps
```
	// Deploy dapp on both chains that use Teleporter Registry
	// Deploy version 2 of Teleporter to both chains
	// Construct AddProtocolVersion txs for both chains
	// Send tx for one of the two chains, verify events emitted
	// Send a message from chain with old version to chain with new version, verify message is received
	// Update minTeleporterVersion on chain with new version
	// Send same message from chain with old version, verify message is not received
	// Send a message from chain with new version to chain with old version, verify message is not received
	// Send tx with new protocol version to other chain, verify events emitted
	// Retry the previously failed message execution, verify message is now able to be delivered to dapp
```
CI

## How is this documented
test comments

### CODE REVIEW

**Comment by minghinmatthewlam** [comment]
The latest `subnet-evm` release does not contain off-chain message changes, so will wait for a new release before merging. Currently running with a separate branch off `master`

File: utils/deployment-utils/deployment_utils.go
Comment (cam-schultz) [inline utils/deployment-utils/deployment_utils.go:192]: I think this would be better in the testing utils, since we wouldn't ever use this function in a real deployment.

File: abi-bindings/go/Teleporter/upgrades/TeleporterRegistry/packing.go
Comment (cam-schultz) [inline abi-bindings/go/Teleporter/upgrades/TeleporterRegistry/packing.go:22]: Is `TeleporterRegistryEntry` an actual type defined in the contract? I think what we're packing here is the packed `ProtocolRegistryEntry` and `DestinationAddress` [here](https://github.com/ava-labs/teleporter/blob/main/contracts/src/Teleporter/upgrades/TeleporterRegistry.sol#L101).

I think we should define the ABI spec for `protocolRegistryEntryType` to be consistent with the actual contract type, then add helpers to extend to `teleporterRegistryEntryType`

File: tests/flows/teleporter_registry.go
Comment (cam-schultz) [inline tests/flows/teleporter_registry.go:25]: We can use the default local network ID defined elsewhere in this PR

File: tests/flows/teleporter_registry.go
Comment (cam-schultz) [inline tests/flows/teleporter_registry.go:134]: Can you add comments to each of these statements mapping them to the test steps described at the beginning of the test function?

File: tests/flows/teleporter_registry.go
Comment (cam-schultz) [inline tests/flows/teleporter_registry.go:29]: IIUC, this test only updates the Teleporter version for subnets A and B. Won't that leave the subnets incompatible with subnet C and the C-Chain for any tests that run after this? 

In any case, can you add a comment to the function description explaining the state changes made by this test? It seems possible that a future test may assume a certain Teleporter address, which would change after this test.

**Review by cam-schultz — COMMENTED** [review]
Overall makes sense to me. Thanks for expanding the utils library as well. Left a handful of relatively minor comments and some questions.

File: tests/testnet/network.go
Comment (minghinmatthewlam) [reply tests/testnet/network.go]: removed

File: tests/flows/teleporter_registry.go
Comment (minghinmatthewlam) [reply tests/flows/teleporter_registry.go]: i'm leaving a constant `defaultNetworkID` field in this test, and left a TODO when we update `avalanche-network-runner` we can directly use the constant from there

File: tests/local/network.go
Comment (minghinmatthewlam) [reply tests/local/network.go]: this is an AVAX amount. Needed an increase because with the new Teleporter Nick's tx we increase the gas price

File: scripts/install_avalanchego_release.sh
Comment (minghinmatthewlam) [reply scripts/install_avalanchego_release.sh]: removed

File: scripts/install_avalanchego_release.sh
Comment (minghinmatthewlam) [reply scripts/install_avalanchego_release.sh:53]: changed it so that it first tries to checkout the tag/version, then if it fails it'll do another attempt by removing hyphen

File: scripts/install_subnetevm_release.sh
Comment (minghinmatthewlam) [reply scripts/install_subnetevm_release.sh:51]: same as above

File: tests/flows/teleporter_registry.go
Comment (minghinmatthewlam) [reply tests/flows/teleporter_registry.go:141]: added

File: abi-bindings/go/Teleporter/upgrades/TeleporterRegistry/packing.go
Comment (minghinmatthewlam) [reply abi-bindings/go/Teleporter/upgrades/TeleporterRegistry/packing.go:18]: updated to match `protocolRegistryEntryType` and chagned function as a helper `PackTeleporterRegistryWarpPayload`

File: tests/flows/teleporter_registry.go
Comment (minghinmatthewlam) [reply tests/flows/teleporter_registry.go:30]: yup that's a good point. Right now after this test runs the C-Chain's `TeleporterRegistry` will be behind on a Teleporter version. The reason tests are still okay for now is because the tests use C-Chain as source mostly, so the destination dapp would still receive from an older version. 

Currently the `subnet-evm` release is not compatible with a `coreth` that has Warp off-chain, will need to update this after so that all subnets match in registered Teleporter versions.

File: tests/flows/teleporter_registry.go
Comment (geoff-vball) [inline tests/flows/teleporter_registry.go:213]: We shouldn't have a single test modifying a global variable like this. We should modify `ConstructKeylessTransaction` to take a gas price and gas limit as params. The current global vars should be set as defaults, and then we can pass different values in from this function.

If we want to be able to call this function multiple times, we can use a variable local to this file to increment as a counter to know what gas price to use.

File: tests/testnet/network.go
Comment (geoff-vball) [inline tests/testnet/network.go:60]: I'm fairly ambivalent, but I think keeping this is a slice and just implementing a getter method on the `testNetwork` struct that takes an `ids.ID` and returns a `SubnetTestInfo` or nil would be slightly cleaner.

File: tests/utils/utils.go
Comment (geoff-vball) [inline tests/utils/utils.go:379]: nit: close parentheses on next line

File: tests/utils/utils.go
Comment (geoff-vball) [inline tests/utils/utils.go:358]: I feel like this function could also call SendTransactionAndWaitForSuccess and return the receipt.

File: tests/utils/utils.go
Comment (geoff-vball) [inline tests/utils/utils.go:889]: nit: I think it's good practice that anytime you use a random number in a test, to print that random number to the screen, in case some random inputs are causing the test to fail, otherwise it might be very hard to track down.

Comment (cam-schultz) [reply tests/utils/utils.go:889]: I'd say we should avoid random numbers altogether. AFAIK there's no need for these messages to be unique, so it might be acceptable to simply send the same value every time. If not, we should derive the value deterministically.

File: tests/utils/utils.go
Comment (minghinmatthewlam) [reply tests/utils/utils.go:358]: going to leave this, similar to `CreateReceiveCrossChainMessageTransaction`, `CreateSendCrossChainMessageTransaction`, etc.

File: scripts/install_avalanchego_release.sh
Comment (minghinmatthewlam) [reply scripts/install_avalanchego_release.sh]: updated

File: tests/utils/utils.go
Comment (minghinmatthewlam) [reply tests/utils/utils.go]: removed using random number, and instead pass in a message

File: tests/flows/teleporter_registry.go
Comment (minghinmatthewlam) [reply tests/flows/teleporter_registry.go:30]: I like that we use the same registry and have random order of e2e tests, so we can that the after updating the registry across all chains, the other e2e tests all still work. Will just need to add the new protocol version to C-chain after a new compatible release is made.

File: tests/flows/teleporter_registry.go
Comment (minghinmatthewlam) [reply tests/flows/teleporter_registry.go]: updated to take in a paramter

File: scripts/install_avalanchego_release.sh
Comment (geoff-vball) [inline scripts/install_avalanchego_release.sh:19]: nit: you've got a second `#` on this line

File: utils/deployment-utils/deployment_utils.go
Comment (geoff-vball) [inline utils/deployment-utils/deployment_utils.go:194]: Can we copy the big.Int here? Returning the pointer to the global var means it can be altered

**Review by geoff-vball — COMMENTED** [review]
Two nits, otherwise lgtm

**Review by geoff-vball — APPROVED** [review]

File: tests/utils/utils.go
Comment (cam-schultz) [inline tests/utils/utils.go:347]: How'd you arrive at 500_000? Curious because I'm implementing similar logic in the relayer to handle off chain messages.

Comment (minghinmatthewlam) [reply tests/utils/utils.go:347]: Initially had the e2e test with too little gas, and in the error message it told me the expected gas

**Review by cam-schultz — APPROVED** [review]
LGTM

**Review by geoff-vball — COMMENTED** [review]
🚀

**Review by geoff-vball — APPROVED** [review]

**Pull request merged**
